### ISSUE TITLE
504 gateway errors on slow connections

### ISSUE THREAD

**Issue by joshuacox** [octohost/octohost#119]
I have one client who is getting 504 errors where no one else is getting them.  Weird thing is I went to their house and got the same error.  I have not seen the error since, but they continue to be plagued by them.  Their connection is tad high on ping, but throughput is nice once it gets going.  At first I thought it was a caching error, but after seeing it on my own hardware I'm wondering what could be triggering this?

**Comment by darron** [comment]
Hmm - it could be that they're timing out on some request that's not quite making it through in time.

I have seen this before, but only once. It was solved by some proxy specific stuff. I think it was something like these values:

```
proxy_connect_timeout       600;
proxy_send_timeout          600;
proxy_read_timeout          600;
send_timeout                600;
```

**Comment by joshuacox** [comment]
Thanks for the quick reply!

According to this page:
http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_read_timeout

75 seconds is the max at least for the connect_timeout, so I'm merely  setting all those values to 75s and trying from there.  I'll report back.  If successful, should I submit a patch for the nginx templates?

**Comment by darron** [comment]
And the default is set at 60s - so that's not much time - let's see if it works and go from there.

I have seen some other timeouts that were caused by uploading larger files that took longer than 60 seconds. See if this helps!

**Comment by joshuacox** [comment]
well and the super weird thing is these timeout are happening in like ten seconds.  So 75s should be more than enough time.  I'm not certain it's not the container itself, I've got a fork of @sameersbn 's 

https://github.com/sameersbn/docker-redmine

it might be even be unicorn borking out somehow on it's end.  But I'm still uncertain as to why it would only be exposed in that one location.

**Comment by joshuacox** [comment]
ya from unicorn's website:

"Slow clients should only be served by placing a reverse proxy capable of fully buffering both the the request and response in between Unicorn and slow clients."

furthermore they have an nginx example:
http://unicorn.bogomips.org/examples/nginx.conf

which begs a couple of questions: if I want to override the templating for a site do I edit
`/etc/nginx/containers/SITE.conf`
or
`/etc/nginx/templates/SITE.ctmpl`

and what's the best way to 'reset'?  I've been doing something like:
```
rm /etc/nginx/*/SITE.*
octo restart SITE
```

I'm starting to think this might be worthy of a Dockerfile magic comment, or another config variable.

**Comment by darron** [comment]
If you want to edit the template, edit the config file in the `/etc/nginx/templates/` file.

If you do an `octo remove site` it will remove the Consul watch, container, src and nginx config: https://github.com/octohost/octohost/blob/master/bin/octo#L619-L637

This does leave the templates around in `/etc/nginx/templates/`.

You can remove all config - including all templates: https://github.com/octohost/octohost/blob/master/bin/octo#L639-L661

Any of that what you're after?

**Comment by joshuacox** [comment]
yep that pretty much sums it up.   So looking at the example config file the fail_timeout sounds very much like it might be the culprit of my woes.  From this page:

http://nginx.org/en/docs/http/ngx_http_upstream_module.html
`By default, the parameter [fail_timeout] is set to 10 seconds.`
which is exactly the amount of time I'm seeing the failure happen at.

So I'm still thinking a mechanism for specifying an alternate template file might be a worthy addition here.  Perhaps include a template.ctmpl in the repo and add a magic comment to the Dockerfile, or another consul config options like the SSL force.   

And/Or a fail_timeout = 0 might be useful across the board, I'm going to experiment with this further.  I'll include the relevant section of the example nginx config, as I find it interesting.

```
  # this can be any application server, not just Unicorn/Rainbows!
  upstream app_server {
    # fail_timeout=0 means we always retry an upstream even if it failed
    # to return a good HTTP response (in case the Unicorn master nukes a
    # single worker for timing out).

    # for UNIX domain socket setups:
    server unix:/path/to/.unicorn.sock fail_timeout=0;

    # for TCP setups, point these to your backend servers
    # server 192.168.0.7:8080 fail_timeout=0;
    # server 192.168.0.8:8080 fail_timeout=0;
    # server 192.168.0.9:8080 fail_timeout=0;
  }
```

**Comment by darron** [comment]
That looks more likely to be the culprit - let me know what you find out.

**Comment by joshuacox** [comment]
Ok so after some testing I think that the fail_timeout=0 is pretty safe, and moreover appears has solved my 504 errors from that one location.  I think it's worth a pull request and make it default. @darron what do you think?

**Comment by joshuacox** [comment]
I've been thinking about this abit more and I think setting fail_timeout=30s max_fails=3 or something like that might be more prudent.  I'm going to test this out a bit, and reply back here.  But to be fair, I've  had no 504 errors from that one source since I've implemented the fail timeout zero.  Now let's see what happens.

**Comment by darron** [comment]
Any word on this?
